### ISSUE TITLE
iOS and Android apps: "select function" does not seem to work in calc

### ISSUE THREAD

**Issue by nikslor** [CollaboraOnline/online#590]
**Describe the bug**
There's a button to choose some often used (I think) functions left to the input bar in calc. The popup that shows up after taping on it seems to be empty - see the screenshots of Online/iOS provided below.

**To Reproduce**
1. Open a spreadsheet file using the iOS app
2. Tap on the "sigma button" left to the input bar

**Expected behavior**
A popup with the functions sum, average, min, max, count should show up.

**Actual behavior**
An empty popup shows up.

**Screenshots**
How this works in normale "Online":
![image](https://user-images.githubusercontent.com/1211300/98676243-a01a3400-235b-11eb-9caa-decedda32cd6.png)

How it looks like on iOS:
<img width="1112" alt="empty_function_popups" src="https://user-images.githubusercontent.com/1211300/98676637-259de400-235c-11eb-9e6c-cb953331c347.png">

**Tablet**
 - OS: iOS
 - Version: 6.4.0 (10)

**Comment by thebearon** [comment]
Indeed, confirmed. Only the iOS app is affected, the dropdown shows fine in the web-based version.

**Comment by tml1024** [comment]
Looking at this now.

**Comment by tml1024** [comment]
Note to self: The list of functions that should be displayed in that tunnelled menu come from sc/uiconfig/scalc/ui/autosum.ui , and the menu is constructed in IMPL_LINK_NOARG(ScInputWindow, DropdownClickHdl, ToolBox *, void) in sc/source/ui/app/inputwin.cxx .

**Comment by tml1024** [comment]
Further notes to self: Possibly the problem is caused by nMaxHeight being -3 at  PopupMenu::ImplExecute(VclPtr<vcl::Window> const&, tools::Rectangle const&, FloatWinPopupFlags, Menu*, bool) at /Volumes/TML13/lo/cp-6.4-ios-debug/vcl/source/window/menu.cxx:2949

**Comment by tml1024** [comment]
Apparently at least some of the 

warn:vcl.quartz:49976:6330278:vcl/quartz/salgdi.cxx:376: AquaSalGraphics::DrawTextLayout() without context

warnings are related to this, drawing the menu entries for the pop-up menu with Sum, Average, etc fails because there is no "context", whatever that then means here. Some fundamental problem in how this stuff with virtual devices and whatnot is handled on iOS.

**Comment by tml1024** [comment]
Progress! Now the menu shows up at presumably correct size, but without entries. 
![IMG_0096](https://user-images.githubusercontent.com/310731/109841829-1ff14480-7c52-11eb-92a2-3cd4d8fe38bf.PNG)

**Comment by thebearon** [comment]
For the record, it's the same in the Android app as in the iOS app: empy dropdown.

In web-based online on a tablet:
- on Android the dropdown is shown, but tapping into it does nothing,
- on iOS it works.

A mildly annoying behavior is that when the dropdown is open, tapping away into the spreadsheet doesn't close it, this is the same on desktop as well (just noting).

**Comment by tml1024** [comment]
Note to self: Work in progress debug diff, stored here for safety:

```
diff --git a/vcl/headless/svpframe.cxx b/vcl/headless/svpframe.cxx
index 20876ef2ff0a..2703603cc738 100644
--- a/vcl/headless/svpframe.cxx
+++ b/vcl/headless/svpframe.cxx
@@ -143,7 +143,16 @@ void SvpSalFrame::LoseFocus()
 SalGraphics* SvpSalFrame::AcquireGraphics()
 {
     SvpSalGraphics* pGraphics = new SvpSalGraphics();
-#ifndef IOS
+#ifdef IOS
+    CGContextRef pCGContext = CGBitmapContextCreate(nullptr, maGeometry.nWidth, maGeometry.nHeight, 8,
+                                                    maGeometry.nWidth * 4, CGColorSpaceCreateDeviceRGB(),
+                                                    kCGImageAlphaPremultipliedFirst | kCGImageByteOrder32Little);
+    SAL_DEBUG("SvpSalFrame::AcquireGraphics() created bitmap context " << pCGContext);
+    CGContextTranslateCTM(pCGContext, 0, maGeometry.nHeight);
+    CGContextScaleCTM(pCGContext, 1, -1);
+    CGLayerHolder aCGLayerHolder(CGLayerCreateWithContext(pCGContext, CGSizeMake(maGeometry.nWidth, maGeometry.nHeight), nullptr));
+    pGraphics->SetVirDevGraphics(aCGLayerHolder, pCGContext);
+#else
     pGraphics->setSurface(m_pSurface, basegfx::B2IVector(maGeometry.nWidth, maGeometry.nHeight));
 #endif
     m_aGraphics.push_back( pGraphics );
diff --git a/vcl/inc/ios/iosinst.hxx b/vcl/inc/ios/iosinst.hxx
index a937fb25fa55..07d69e4b06bc 100644
--- a/vcl/inc/ios/iosinst.hxx
+++ b/vcl/inc/ios/iosinst.hxx
@@ -41,7 +41,6 @@ public:
 
     css::uno::Reference< css::uno::XInterface > CreateClipboard( const css::uno::Sequence< css::uno::Any >& i_rArguments ) override;
 
-    void GetWorkArea( tools::Rectangle& rRect );
     SalFrame* CreateFrame( SalFrame* pParent, SalFrameStyleFlags nStyle ) override;
     SalFrame* CreateChildFrame( SystemParentData* pParent, SalFrameStyleFlags nStyle ) override;
 };
diff --git a/vcl/inc/quartz/CGHelpers.hxx b/vcl/inc/quartz/CGHelpers.hxx
index c83477ead4fe..86132dbe0bca 100644
--- a/vcl/inc/quartz/CGHelpers.hxx
+++ b/vcl/inc/quartz/CGHelpers.hxx
@@ -95,9 +95,17 @@ public:
 
     void set(CGContextRef const& pContext) { mpContext = pContext; }
 
-    void saveState() { CGContextSaveGState(mpContext); }
+    void saveState()
+    {
+        assert(mpContext != nullptr);
+        CGContextSaveGState(mpContext);
+    }
 
-    void restoreState() { CGContextRestoreGState(mpContext); }
+    void restoreState()
+    {
+        assert(mpContext != nullptr);
+        CGContextRestoreGState(mpContext);
+    }
 };
 
 #endif // INCLUDED_VCL_INC_QUARTZ_CGHELPER_HXX
diff --git a/vcl/inc/quartz/salgdi.h b/vcl/inc/quartz/salgdi.h
index aac4e140f5f9..b0299e749bd5 100644
--- a/vcl/inc/quartz/salgdi.h
+++ b/vcl/inc/quartz/salgdi.h
@@ -205,7 +205,6 @@ public:
     void                    ImplDrawPixel( long nX, long nY, const RGBAColor& ); // helper to draw single pixels
 
     bool                    CheckContext();
-    CGContextRef            GetContext();
 #ifdef MACOSX
     void                    UpdateWindow( NSRect& ); // delivered in NSView coordinates
     void                    RefreshRect( const NSRect& );
diff --git a/vcl/ios/iosinst.cxx b/vcl/ios/iosinst.cxx
index 65963ef9c099..f3972cff1985 100644
--- a/vcl/ios/iosinst.cxx
+++ b/vcl/ios/iosinst.cxx
@@ -42,12 +42,6 @@ public:
     virtual bool ErrorTrapPop( bool ) { return false; }
 };
 
-void IosSalInstance::GetWorkArea( tools::Rectangle& rRect )
-{
-    rRect = tools::Rectangle( Point( 0, 0 ),
-                       Size( viewWidth, viewHeight ) );
-}
-
 IosSalInstance *IosSalInstance::getInstance()
 {
     if (!ImplGetSVData())
@@ -93,11 +87,6 @@ public:
             SetPosSize(0, 0, viewWidth, viewHeight, SAL_FRAME_POSSIZE_WIDTH | SAL_FRAME_POSSIZE_HEIGHT);
     }
 
-    virtual void GetWorkArea( tools::Rectangle& rRect ) override
-    {
-        IosSalInstance::getInstance()->GetWorkArea( rRect );
-    }
-
     virtual void ShowFullScreen( bool, sal_Int32 ) override
     {
         SetPosSize( 0, 0, viewWidth, viewHeight,
diff --git a/vcl/ios/salios.cxx b/vcl/ios/salios.cxx
index fc972909cdd2..1ac3ac15c60e 100644
--- a/vcl/ios/salios.cxx
+++ b/vcl/ios/salios.cxx
@@ -278,19 +278,18 @@ void AquaSalGraphics::copyArea( long nDstX, long nDstY,long nSrcX, long nSrcY,
 void AquaSalGraphics::SetVirDevGraphics(CGLayerHolder const & rLayer, CGContextRef xContext,
                                         int nBitmapDepth)
 {
-    SAL_INFO( "vcl.quartz", "SetVirDevGraphics() this=" << this << " layer=" << rLayer.get() << " context=" << xContext );
+    SAL_DEBUG( "AquaSalGraphics::SetVirDevGraphics() this=" << this << " layer=" << rLayer.get() << " context=" << xContext );
 
     mbPrinter = false;
     mbVirDev = true;
 
     // set graphics properties
     maLayer = rLayer;
+    SAL_DEBUG("================ Setting maContextHolder to " << xContext);
     maContextHolder.set(xContext);
 
     mnBitmapDepth = nBitmapDepth;
 
-    mbForeignContext = xContext != NULL;
-
     // return early if the virdev is being destroyed
     if( !xContext )
         return;
@@ -319,6 +318,7 @@ void AquaSalGraphics::SetVirDevGraphics(CGLayerHolder const & rLayer, CGContextR
         mpXorEmulation->SetTarget(mnWidth, mnHeight, mnBitmapDepth, maContextHolder.get(), maLayer.get());
         if( mpXorEmulation->IsEnabled() )
         {
+            SAL_DEBUG("================ Setting maContextHolder to mpXorEmulation->GetMaskContext()=" << mpXorEmulation->GetMaskContext());
             maContextHolder.set(mpXorEmulation->GetMaskContext());
         }
     }
@@ -347,6 +347,7 @@ void AquaSalVirtualDevice::Destroy()
         {
             mpGraphics->SetVirDevGraphics(nullptr, nullptr);
         }
+        SAL_DEBUG("AquaSalVirtualDevice::Destroy(): Releasing layer " << maLayer.get());
         CGLayerRelease(maLayer.get());
         maLayer.set(nullptr);
     }
diff --git a/vcl/osx/salgdiutils.cxx b/vcl/osx/salgdiutils.cxx
index abd90deaa84b..0170bc6edefa 100644
--- a/vcl/osx/salgdiutils.cxx
+++ b/vcl/osx/salgdiutils.cxx
@@ -204,15 +204,6 @@ bool AquaSalGraphics::CheckContext()
     return maContextHolder.isSet();
 }
 
-CGContextRef AquaSalGraphics::GetContext()
-{
-    if (!maContextHolder.isSet())
-    {
-        CheckContext();
-    }
-    return maContextHolder.get();
-}
-
 /**
  * Blit the contents of our internal maLayer state to the
  * associated window, if any; cf. drawRect event handling
diff --git a/vcl/quartz/salgdi.cxx b/vcl/quartz/salgdi.cxx
index db272df57587..1e8d21b52be2 100644
--- a/vcl/quartz/salgdi.cxx
+++ b/vcl/quartz/salgdi.cxx
@@ -212,7 +212,7 @@ AquaSalGraphics::AquaSalGraphics()
     , mbForeignContext( false )
 #endif
 {
-    SAL_INFO( "vcl.quartz", "AquaSalGraphics::AquaSalGraphics() this=" << this );
+    SAL_DEBUG( /* "vcl.quartz", */ "AquaSalGraphics::AquaSalGraphics() this=" << this );
 
     for (int i = 0; i < MAX_FALLBACK; ++i)
         mpTextStyle[i] = nullptr;
@@ -223,7 +223,7 @@ AquaSalGraphics::AquaSalGraphics()
 
 AquaSalGraphics::~AquaSalGraphics()
 {
-    SAL_INFO( "vcl.quartz", "AquaSalGraphics::~AquaSalGraphics() this=" << this );
+    SAL_DEBUG( "AquaSalGraphics::~AquaSalGraphics() this=" << this );
 
     if( mxClipPath )
     {
@@ -241,6 +241,7 @@ AquaSalGraphics::~AquaSalGraphics()
 #endif
     if (maLayer.isSet())
     {
+        SAL_DEBUG("AquaSalGraphics::~AquaSalGraphics(): releasing " << maLayer.get());
         CGLayerRelease(maLayer.get());
     }
     else if (maContextHolder.isSet()
@@ -251,6 +252,7 @@ AquaSalGraphics::~AquaSalGraphics()
     {
         // destroy backbuffer bitmap context that we created ourself
         CGContextRelease(maContextHolder.get());
+        SAL_DEBUG("================ Setting maContextHolder to null");
         maContextHolder.set(nullptr);
     }
 }
@@ -370,14 +372,6 @@ bool AquaSalGraphics::AddTempDevFont( PhysicalFontCollection*,
 
 void AquaSalGraphics::DrawTextLayout(const GenericSalLayout& rLayout)
 {
-#ifdef IOS
-    if (!CheckContext())
-    {
-        SAL_WARN("vcl.quartz", "AquaSalGraphics::DrawTextLayout() without context");
-        return;
-    }
-#endif
-
     const CoreTextStyle& rStyle = *static_cast<const CoreTextStyle*>(&rLayout.GetFont());
     const FontSelectPattern& rFontSelect = rStyle.GetFontSelectPattern();
     if (rFontSelect.mnHeight == 0)
@@ -865,22 +859,7 @@ void AquaSalGraphics::RefreshRect(float lX, float lY, float lWidth, float lHeigh
 
 bool AquaSalGraphics::CheckContext()
 {
-    if (mbForeignContext)
-    {
-        SAL_INFO("vcl.ios", "CheckContext() this=" << this << ", mbForeignContext, return true");
-        return true;
-    }
-
-    SAL_INFO( "vcl.ios", "CheckContext() this=" << this << ",  not foreign, return false");
-    return false;
-}
-
-CGContextRef AquaSalGraphics::GetContext()
-{
-    if (!maContextHolder.isSet())
-        CheckContext();
-
-    return maContextHolder.get();
+    return true;
 }
 
 #endif
diff --git a/vcl/quartz/salvd.cxx b/vcl/quartz/salvd.cxx
index e9fe55d27406..9135bfe96d5a 100644
--- a/vcl/quartz/salvd.cxx
+++ b/vcl/quartz/salvd.cxx
@@ -127,6 +127,18 @@ AquaSalVirtualDevice::AquaSalVirtualDevice( AquaSalGraphics* pGraphic, long &nDX
                 mnBitmapDepth = 0;
                 break;
         }
+
+#ifdef IOS
+        CGContextRef pCGContext = CGBitmapContextCreate(nullptr, nDX, nDY, 8,
+                                                        nDX * 4, CGColorSpaceCreateDeviceRGB(),
+                                                        kCGImageAlphaPremultipliedFirst | kCGImageByteOrder32Little);
+        SAL_DEBUG("AquaSalVirtualDevice::AquaSalVirtualDevice() created bitmap context " << pCGContext);
+        CGContextTranslateCTM(pCGContext, 0, nDY);
+        CGContextScaleCTM(pCGContext, 1, -1);
+        maLayer.set(CGLayerCreateWithContext(pCGContext, CGSizeMake(nDX, nDY), nullptr));
+        mpGraphics->SetVirDevGraphics(maLayer, pCGContext);
+#endif
+
 #ifdef MACOSX
         // inherit resolution from reference device
         if( pGraphic )
```

**Comment by nikslor** [comment]
This is fixed in 21.6.1 (1). The position of the popup is not perfect, but good enough. I think we can close this issue.
